src/WebJobs.Extensions.DurableTask/DurableTaskExtension.cs

```diff
@@ -399,9 +399,19 @@ internal void DeregisterOrchestrator(FunctionName orchestratorFunction)
 
         internal void RegisterActivity(FunctionName activityFunction, ITriggeredFunctionExecutor executor)
         {
-            if (!this.registeredActivities.TryAdd(activityFunction, executor))
+            if (this.registeredActivities.ContainsKey(activityFunction))
             {
-                throw new ArgumentException($"The activity function named '{activityFunction}' is already registered.");
+                if (!this.registeredActivities.TryUpdate(activityFunction, executor, null))
+                {
+                    throw new ArgumentException($"The executor of an activity function named '{activityFunction}' cannot be updated from previous null value.");
+                }
+            }
+            else
+            {
+                if (!this.registeredActivities.TryAdd(activityFunction, executor))
+                {
+                    throw new ArgumentException($"The activity function named '{activityFunction}' is already registered.");
+                }

```

Comment:
I like this approach and how it ensures correctness. Maybe we can simplify it slightly?

```csharp
// Allow adding with a null key and subsequently updating with a non-null key.
if (!this.registeredActivities.TryUpdate(activityFunction, executor, null))
{
    if (!this.registeredActivities.TryAdd(activityFunction, executor))
    {
        throw new ArgumentException("The activity function named '{activityFunction}' is already fully registered."};
    }
}
```

Reply:
Great optimization! Thank you very much!